reps/2024-05-21-kuberay-authentication.md

```diff
@@ -0,0 +1,157 @@
+3. A ServiceAccount is created per RayCluster resource
+
+The sidecar ensures only requests with the correct authorization tokens can access the Ray head. More details on how tokens are authenticated below.
+For the MVP, the sidecar will be implemented using Envoy configured with an [External Authorizer](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/ext_authz_filter).
+
+## Access Control with TokenReview API
+
+The sidecar authenticator will use the TokenReview API to authenticate tokens passed into the authorization headers of each request.
+Tokens can be created from Kubernetes ServiceAccounts or an external identity provider as long as the Kubernetes API Server is configured with an external authorizer.
+Each RayCluster will be provisioned a default ServiceAccount that KubeRay Operator will use when authenticating to the Ray head (specifically for RayJob and RayService).
+Users can use the default ServiceAccount in the absence of external identity providers.
+
+Example of a TokenReview request that contains a token:
+```
+apiVersion: authentication.k8s.io/v1
+kind: TokenReview
+spec:
+  token: <your_token_here>
+```
+
+Example of a TokenReview response with user information about the token:
+```
+apiVersion: authentication.k8s.io/v1
+kind: TokenReview
+status:
+  authenticated: true
+  user:
+    username: system:serviceaccount:default:ray-cluster
+    uid: <some_uid>
+    groups: ["system:serviceaccounts", "system:serviceaccounts:default"]
+```
+
+## User to Ray Head Authentication
+
+When submitting jobs using the Ray job CLI:
+```
+export RAY_JOB_HEADERS="Authorization: Bearer $(kubectl create token ray-cluster)"
+OR
+export RAY_JOB_HEADERS="Authorization: Bearer $(gcloud auth print-access-token)" # example external identity provider
+
+ray job submit...
+```
+
+When using Ray interactive client:
+```
+export RAY_CLUSTER_TOKEN=$(kubectl create token ray-cluster)
+OR
+export RAY_CLUSTER_TOKEN=$(gcloud auth print-access-token) # example external identity provider
+------------------------------------------------------------
+
+import os
+import ray
+import base64
+
+def get_metadata():
+    """
+    Get authentication header metadata for ray.util.connect
+    """
+    headers = {"Authorization": f"Bearer {os.environ["RAY_CLUSTER_TOKEN"]}"}
+    return [(key.lower(), value) for key, value in headers.items()]
+
+ray.init(
+    "ray-cluster:443",
+    metadata=get_metadata(),
+    secure=True,
+)
+```
+
+## RayCluster API Changes
+
+Enabling authentication should be optional per RayCluster. Two new fields `enableAuthentication` and `authenticationOptions` will be added to the RayCluster spec to enable this capability and configure allowed principals.
+If no allowed principals are specified, it will default to a ServiceAccount with the same namespace and name as the RayCluster.
+
+```
+apiVersion: ray.io/v1
+kind: RayCluster
+metdata:
+  ...
+spec:
+  enableAuthentication: true
+  authenticationOptions:
```

Comment:
In some organisations, RayCluster resources will be created by the "data scientist" / end-user persona, while the security aspect will be owned by the platform administrator. The current proposal makes it difficult to enforce that model with Kubernetes RBAC.

The API surface of that new feature should ideally follow usual API graduation. Including it into the RayCluster API, that's v1, will make it difficult to iterate on it and make it evolve, by complying to v1 guarantees.

The RayCluster API has already a large surface. Could the proposal explore the solution space, and research a more modular approach for the configurability aspect, that'd be more symmetrical to the sidecar implementation proposal, i.e. non intrusive, and generally try to keep the single-responsibility principle for the RayCluster API as much as possible.

Reply:
> In some organisations, RayCluster resources will be created by the "data scientist" / end-user persona, while the security aspect will be owned by the platform administrator. The current proposal makes it difficult to enforce that model with Kubernetes RBAC.

My assumption is that a platform admin can grant role / rolebindings that give data scientists namespace-wide access to RayClusters (both creating them and accessing them) so they don't need to manage security on a per-cluster basis. You can also make this a ClusterRole to give access to all RayClusters in the cluster. In organizations where data scientists also create RayClusters, they also have the option to by-pass the RBAC check with the fixed list of allowed users (`allowedPrincipals`).

> The API surface of that new feature should ideally follow usual API graduation. Including it into the RayCluster API, that's v1, will make it difficult to iterate on it and make it evolve, by complying to v1 guarantees.

I think this is more of a concern around feature gates and gating changes to v1 APIs that might not be mature yet. I will try to explore adding Kubernetes-like feature gates to KubeRay as we implement this. FYI @kevin85421 

> The RayCluster API has already a large surface. Could the proposal explore the solution space, and research a more modular approach for the configurability aspect, that'd be more symmetrical to the sidecar implementation proposal, i.e. non intrusive, and generally try to keep the single-responsibility principle for the RayCluster API as much as possible.

Let me give this more thought and let's revisit during code review? In theory the addition of a sidecar is already modular because RayCluster lets you override the entire pod template. So my initial thinking is that enabling authentication in KubeRay should default to a solution that will work for most users out of the box (with something like kube-rbac-proxy)

Reply:
We can discuss more during code review about the API.